src/lib.rs

```diff
@@ -2,269 +2,186 @@
+        handle_error(archive_write_free(aw), aw)?;
     }
 
-    #[test]
-    fn uncompress_tar_lzma() {
-        let (_dir_handle, calls) = create_echo_bins(&["cat", "tar", "lzcat"]).unwrap();
-        uncompress("test.tar.lzma", &PathBuf::from("target"), Kind::TarLZMA)
-            .expect("Failed to uncompress file");
-        assert_calls(
-            &calls,
-            &[
-                "cat test.tar.lzma",
-                "lzcat",
-                "tar --same-owner --preserve-permissions --xattrs -xC target",
-            ],
-        );
-    }
-
-    #[test]
-    fn uncompress_tar_lzip() {
-        let (_dir_handle, calls) = create_echo_bins(&["cat", "tar", "lzip"]).unwrap();
-        uncompress("test.tar.lzip", &PathBuf::from("target"), Kind::TarLZip)
-            .expect("Failed to uncompress file");
-        assert_calls(
-            &calls,
-            &[
-                "cat test.tar.lzip",
-                "lzip -dc",
-                "tar --same-owner --preserve-permissions --xattrs -xC target",
-            ],
-        );
-    }
-
-    #[test]
-    fn uncompress_tar() {
-        let (_dir_handle, calls) = create_echo_bins(&["cat", "tar"]).unwrap();
-        uncompress("test.tar", &PathBuf::from("target"), Kind::Tar)
-            .expect("Failed to uncompress file");
-        assert_calls(
-            &calls,
-            &[
-                "cat test.tar",
-                "tar --same-owner --preserve-permissions --xattrs -xC target",
-            ],
-        );
-    }
-
-    #[test]
-    fn uncompress_gz() {
-        let (dir_handle, calls) = create_echo_bins(&["cat", "zcat"]).unwrap();
-        uncompress("test.gz", &dir_handle.path().join("target"), Kind::GZip)
-            .expect("Failed to uncompress file");
-        assert_calls(&calls, &["cat test.gz", "zcat"]);
-    }
-
-    #[test]
-    fn uncompress_bz2() {
-        let (dir_handle, calls) = create_echo_bins(&["cat", "bzcat"]).unwrap();
-        uncompress("test.bz2", &dir_handle.path().join("target"), Kind::BZip2)
-            .expect("Failed to uncompress file");
-        assert_calls(&calls, &["cat test.bz2", "bzcat"]);
-    }
-
-    #[test]
-    fn uncompress_xz() {
-        let (dir_handle, calls) = create_echo_bins(&["cat", "xzcat"]).unwrap();
-        uncompress("test.xz", &dir_handle.path().join("target"), Kind::Xz)
-            .expect("Failed to uncompress file");
-        assert_calls(&calls, &["cat test.xz", "xzcat"]);
-    }
+    Ok(())
+}
 
-    #[test]
-    fn uncompress_lzma() {
-        let (dir_handle, calls) = create_echo_bins(&["cat", "lzcat"]).unwrap();
-        uncompress("test.lzma", &dir_handle.path().join("target"), Kind::LZMA)
-            .expect("Failed to uncompress file");
-        assert_calls(&calls, &["cat test.lzma", "lzcat"]);
-    }
+// reads the input buffer and extract to another buffer.
+pub fn buffered_uncompress(source: &[u8], target: &mut [u8]) -> compress::Result<()> {
+    init_archive(|mut entry, ar, aw| unsafe {
```

Comment:
I think the name `init_archive` is not really meaningful here. I suggest something like `run_with_archive`.